Custom compare support for the If methods

007hacky007 opened the pull request:
Added support for custom comparison operation and target. 

I tried to implement this same way as it is implemented the original Golang library.

Unfortunately I had to change ClientInterface as well. It could be implemented with adding new methods like 'putIf2' and 'deleteIf2' to the interface, but it would be dirty and not worth it in my opinion.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aternosorg/php-etcd?pullRequest=1) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aternosorg/php-etcd?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aternosorg/php-etcd?pullRequest=1) it.</sub>

007hacky007 commented:
I'm going to change parameters positioning. I wanted all compare related parameters together, but I guess it's better to retain compatibility. 

As for the `$compareOp`, I can indeed change it to directly use `CompareResult`, however I wanted to copy original Golang implementation of the etcd - check https://github.com/etcd-io/etcd/blob/master/clientv3/compare.go#L33 as it would be nearest to what etcd authors designed in my opinion.

matthi4s commented:
Hmm, I get that but I don't really know why it's that way in Golang. Using the constants directly needs no further validation and also requires no changes if new constants are added or changed. Using constants over strings also provides better IDE support.

007hacky007 commented:
`Compare` class does not accept nothing else than `string`, so for the clarity I've specified the type of the variable you are putting into the `setValue` method of the `Compare` class. 

Also I can change it, so it will allow `bool|string` and defining `false` as well, however it may be confusing what behavior to expect when you set `$compareValue = false` AND at the same time `$compareTarget` to anything else than `CompareTarget::VERSION`, because it is not to be expected that setting `$compareValue = false` actually rewrites `$compareValue` to `(string)0`. 

So if one would call 
`$client->putIf("key", "1", false, false, CompareResult::EQUAL, CompareTarget::VALUE);`

the behavior may be other than expected (what happens with `false`? One may think it changes to `(string)false` or to `""`). It's not clear that you're changing `false` to `"0"` until you read the code. I think it's better to keep `setValue` input as original type, `string` - for the sake of clarity.

matthi4s commented:
Maybe we should go for a public `requestIf` function and public `getCompare` and maybe even `getPutOperation`, `getDeleteOperation` for better customization support. And keep the current `deleteIf` and `putIf` simple.

007hacky007 commented:
All right, I've changed both `putIf` and `deleteIf` methods, so they accept original parameters. 

I've added new methods `getGetOperation`, `getPutOperation` and `getDeleteOperation`. 
`requestIf` method is now public as well as `getCompare` method. 

`requestIf` also now accepts array of `Compare` and `RequestOp` objects, because you can use more comparisons in conjunction in etcd as well as you can perform multiple operations in single transaction. I realize (as I'm writing this) you can actually set multiple `RequestOp` objects for the failure part of the transaction, so this can be added as well.

matthi4s commented on src/Client.php:
```diff
@@ -179,104 +179,145 @@ public function delete(string $key)
+     * @param array $requestOperations operations to perform on success, array of RequestOp objects
+     * @param array|null $failureOperations operations to perform on failure, array of RequestOp objects
+     * @param array $compare array of Compare objects
      * @param bool $returnNewValueOnFail
```
```suggestion
```

matthi4s commented on src/Client.php:
```diff
@@ -179,104 +179,145 @@ public function delete(string $key)
      * @throws InvalidResponseStatusCodeException
      */
-    protected function requestIf(string $key, $previousValue, RequestOp $requestOperation, bool $returnNewValueOnFail = false)
+    public function requestIf(string $key, array $requestOperations, ?array $failureOperations, array $compare): TxnResponse
```
```suggestion
    public function txnRequest(string $key, array $requestOperations, ?array $failureOperations, array $compare): TxnResponse
```

I think the name fits no longer and would suggest `txnRequest` instead.
